### ISSUE TITLE
4.0.0-rc.0 - infinite re-render in list component when useResolvedPath is used

### ISSUE THREAD

**Issue by panfiva** [marmelab/react-admin#7452]
Adding `const resolved = useResolvedPath('')` cases the following issues:
- List component causes infinite re-render.
- web page URL is automatically updated to include URL parameters (when loading a page by link `https://vw0pg1.sse.codesandbox.io/#/posts`, the browser redirects to `https://vw0pg1.sse.codesandbox.io/#/posts?displayedFilters=%7B%7D&filter=%7B%7D&order=ASC&page=1&perPage=10&sort=id`)

**What you were expecting:**
useResolvedPath usage in list component should not impact react-admin behavior:
 - when I enter `https://vw0pg1.sse.codesandbox.io/#/posts` to blank page in browser, it should stay in `https://vw0pg1.sse.codesandbox.io/#/posts` without adding parameters
 - the page should not go into infinite re-render loop

**What happened instead:**

**Steps to reproduce:**

**Related code:**

https://codesandbox.io/s/winter-sky-vw0pg1?file=/src/posts/PostList.tsx:707-1355

Notice how console.log keeps printing messages, and also render count keeps updating.

If I comment out lines that cause issues, then react-admin starts behaving properly

```
export const PostList = (props: any): JSX.Element | null => {
  const ref = React.useRef(1);
  ref.current = ref.current + 1;

  const basename = useBasename();
  const resolvedPath = useResolvedPath(""); // THIS LINE CAUSES ISSUES
  console.log({ basename, resolvedPath });  // THIS LINE CAUSES ISSUES

  const filters = [
    <TextInput
      key="1"
      label="id"
      source="id"
      alwaysOn
      sx={{ width: "100%", minWidth: "350px" }}
    />
  ];

  return (
    <Fragment>
      <div>{ref.current}</div>
      <List {...props} filters={filters}>
        <Datagrid rowClick="edit">
          <TextField source="id" label="ID" />
        </Datagrid>
      </List>
    </Fragment>
  );
};
```

**Other information:**

**Environment**

* React-admin version: 4.0.0-rc.0
* Last version that did not exhibit the issue (if applicable):
* React version:
* Browser:
* Stack trace (in case of a JS error):

**Comment by panfiva** [comment]
I noticed that if I add `disableSyncWithLocation` to my list, then I no longer have this issue. I suspect that the issue is caused by `navigate` call inside `useListParams.ts`; it seems it is trying to navigate to on each re-render, triggering another re-render.

**Comment by panfiva** [comment]
If the issue is with navigate call, react-admin will need to check if current values of parameters match or do not match expected parameters before navigation is issued; the check should also make sure that default values are properly evaluated so that unnecessary re-render is not triggered from `/posts`  to `posts?displayedFilters=%7B%7D&filter=%7B%7D&order=ASC&page=1&perPage=10&sort=id`

**Comment by slax57** [comment]
Hi!
Thank you for submitting this issue.
I can indeed reproduce the problem, and I believe this is due to a recent change we made. I'm working on the fix right now.

In the meantime, please note that you won't have this problem if you declare your `filters` **outside** the `PostList` component.

We'll keep you updated regarding the fix.
Cheers!

**Comment by fzaninotto** [comment]
Fixed by #7457
